### PR TITLE
fix(admin-api/status) removed counters and added simple DB ping

### PR DESCRIPTION
**Pull request by shashiranjan84**
### Summary

In favor to make `/status` admin API fast and constant time execution(short of), counters are removed and DB reachability status added.

### Full changelog

* Added reachability check to DB is done by firing a light query
  to check if keyspace/database exists in DB.
* Removed counters from `/status` api response and `reachable` status added.
image

* Updated related test.

Note: Existing version of DataDog agent will not work with above fix.
  A separate PR would be sent to DD repo to fix that too.

### CODE REVIEW

File: kong/dao/db/cassandra.lua [hunk 1]
```diff
@@ -24,8 +24,8 @@ _M.dao_insert_values = {
     return uuid()
   end,
   timestamp = function()
-    -- return time in UNIT millisecond, and PRECISION millisecond 
-    return math.floor(timestamp.get_utc_ms()) 
+    -- return time in UNIT millisecond, and PRECISION millisecond
+    return math.floor(timestamp.get_utc_ms())
```
Comment (p0pr0ck5) [inline kong/dao/db/cassandra.lua]: can we avoid these whitespace cleanups. glad we're cleaning up, but it dirties the commit history. we can have a separate style clean commit later.

File: kong/dao/db/postgres.lua [hunk 2]
```diff
@@ -497,6 +497,7 @@ function _M:current_migrations()
   end
 end
 
+
```
Comment (p0pr0ck5) [inline kong/dao/db/postgres.lua]: another unnecessary change?

File: kong/api/routes/kong.lua [hunk 3]
```diff
@@ -90,15 +90,17 @@ return {
...
-        if err then
-          return helpers.responses.send_HTTP_INTERNAL_SERVER_ERROR(err)
-        end
-        status_response.database[k] = count
+      -- ping DB
+      local err
+      status_response.database.reachable, err = dao.db:is_reachable()
+
+      if err ~= nil then
+        ngx.log(ngx.ERR, err)
```
Comment (p0pr0ck5) [inline kong/api/routes/kong.lua]: should we report the error, or a summary of the error, as part of the response? my gut feeling is 'no', just bringing this up as a discussion point

**Comment by thefosk** [comment]
Shouldn't this PR also update the plugins that are currently leveraging on the counters?

File: kong/dao/db/cassandra.lua [hunk 4]
```diff
@@ -660,4 +660,36 @@ function _M:record_migration(id, name)
...
+    ]]
+  end
+
+  local rows, err = self:query(q_keyspace_exists, {
+    self.cluster_options.keyspace
+  }, {
+    prepared = false
+  }, nil, true)
+
+  if not rows then return false, err end
```
Comment (p0pr0ck5) [inline kong/dao/db/cassandra.lua]: for style, i think we avoid one-line conditional blocks?

**Comment by shashiranjan84** [comment]
@thefosk no, counter method is still there and being referenced at many places. So it would need separate card/focus.

File: kong/dao/db/cassandra.lua (hunk 1 above)
Comment (shashiranjan84) [inline kong/dao/db/cassandra.lua]: Unintentional fix done by IDE. I'll take care of this next time.

File: kong/dao/db/cassandra.lua (hunk 4 above)
Comment (shashiranjan84) [inline kong/dao/db/cassandra.lua]: Note sure, personally I would prefer indented conditional block, but I Kong is full of one-liner conditional statements.

File: kong/api/routes/kong.lua (hunk 3 above)
Comment (shashiranjan84) [inline kong/api/routes/kong.lua]: I had this question, but I guess it okay to just return status.

File: kong/dao/db/postgres.lua (hunk 2 above)
Comment (shashiranjan84) [inline kong/dao/db/postgres.lua]: Updated.

**Comment by thefosk** [comment]
Don't those agents rely on the counters returned by the `/status` endpoint?

**Comment by shashiranjan84** [comment]
@thefosk sorry, what agent you talking about? Only agent I am aware of is DataDog, which I already put in note.

File: kong/api/routes/kong.lua (hunk 3 above)
Comment (thibaultcha) [inline kong/api/routes/kong.lua]: We need to at least give context to this error:
```
ngx.log(ngx.ERR, "failed to reach database as part of ", 
                 "/status endpoint: ", err)
```

File: kong/dao/db/cassandra.lua (hunk 4 above)
Comment (thibaultcha) [inline kong/dao/db/cassandra.lua]: Those are legacy, please use indented conditional branches indeed.

Comment (thibaultcha) [inline kong/dao/db/cassandra.lua]: Additionally, and by convention, we return `nil, err` on error, and not `false`. This is also valid later down this function.

File: kong/dao/db/postgres.lua [hunk 5]
```diff
@@ -516,4 +516,21 @@ function _M:record_migration(id, name)
   return true
 end
 
+function _M:is_reachable()
+
```
Comment (thibaultcha) [inline kong/dao/db/postgres.lua]: this blank line is undesired

File: kong/dao/db/postgres.lua [hunk 6]
```diff
@@ -516,4 +516,21 @@ function _M:record_migration(id, name)
...
+  local query = fmt(
+    [[
+      SELECT * FROM pg_stat_database where datname='%s'
+    ]], self.query_options.database)
```
Comment (thibaultcha) [inline kong/dao/db/postgres.lua]: I'm not sure we need the multi-line string here:
```lua
local query = fmt("SELECT * FROM pg_stat_database where datname='%s'",
                  self.query_options.database)

```

File: kong/dao/db/postgres.lua [hunk 7]
```diff
@@ -516,4 +516,21 @@ function _M:record_migration(id, name)
...
+
+  local rows, err = self:query(query)
+  if not rows then return false, err end
```
Comment (thibaultcha) [inline kong/dao/db/postgres.lua]: ditto: indented block vs single line and return failure value (same below)

File: kong/api/routes/kong.lua [hunk 8]
```diff
@@ -90,15 +90,17 @@ return {
...
 
-      for k, v in pairs(dao.daos) do
-        local count, err = v:count()
-        if err then
-          return helpers.responses.send_HTTP_INTERNAL_SERVER_ERROR(err)
-        end
-        status_response.database[k] = count
+      -- ping DB
+      local err
+      status_response.database.reachable, err = dao.db:is_reachable()
```
Comment (thibaultcha) [inline kong/api/routes/kong.lua]: Due to the below changes with the return value (and because this `reachable()` function might be used elsewhere and should not be tied to this endpoint for the sake of modularisation), this could become:
```
local ok, err = dao.db:reachable()
if not ok then
  ngx.log(ngx.ERR, "msg...", err)

else
  status_response.database.reachable = true
end
```

File: kong/dao/db/cassandra.lua [hunk 9]
```diff
@@ -660,4 +660,36 @@ function _M:record_migration(id, name)
   return true
 end
 
+function _M:is_reachable()
```
Comment (thibaultcha) [inline kong/dao/db/cassandra.lua]: We prefer less cumber-stone names, here, `reachable()` should better suited.

**Comment by shashiranjan84** [comment]
Updated as advised.

**Comment by thibaultcha** [comment]
LGTM, I'll merge manually with some minor leftover edits

**Comment by shashiranjan84** [comment]
Thanks @thibaultcha.

**Comment by thibaultcha** [comment]
On second thought, I do think the logic here is not totally aligned with what has been discussed... See #2567.

**Comment by thibaultcha** [comment]
Done with #2567
